pallets/orml-tokens-extension/src/lib.rs

```diff
@@ -0,0 +1,293 @@
+
+pub mod default_weights;
+
+#[cfg(test)]
+mod tests;
+
+mod ext;
+
+mod types;
+
+pub use pallet::*;
+
+pub(crate) type BalanceOf<T> = <<T as orml_currencies::Config>::MultiCurrency as MultiCurrency<
+	<T as frame_system::Config>::AccountId,
+>>::Balance;
+
+pub(crate) type CurrencyOf<T> = <<T as orml_currencies::Config>::MultiCurrency as MultiCurrency<
+	<T as frame_system::Config>::AccountId,
+>>::CurrencyId;
+
+pub(crate) type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use crate::types::CurrencyDetails;
+	use frame_support::{pallet_prelude::*, transactional};
+	use frame_system::{ensure_signed, pallet_prelude::OriginFor};
+
+	/// ## Configuration
+	/// The pallet's configuration trait.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + orml_tokens::Config + orml_currencies::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Weight information for the extrinsics in this module.
+		type WeightInfo: WeightInfo;
+
+		/// Type that allows for checking if currency type is ownable by users
+		type CurrencyIdChecker: CurrencyIdCheck<CurrencyId = CurrencyOf<Self>>;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn currency_details)]
+	pub type CurrencyData<T: Config> =
+		StorageMap<_, Blake2_128Concat, CurrencyOf<T>, CurrencyDetails<AccountIdOf<T>>>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Some currency was issued.
+		Mint { currency_id: CurrencyOf<T>, to: AccountIdOf<T>, amount: BalanceOf<T> },
+		/// Some currency was burned.
+		Burned { currency_id: CurrencyOf<T>, from: AccountIdOf<T>, amount: BalanceOf<T> },
+		/// Some currency class was created.
+		Created { currency_id: CurrencyOf<T>, creator: AccountIdOf<T>, owner: AccountIdOf<T> },
+		/// Some currency was destroyed (it's data)
+		Destroyed { currency_id: CurrencyOf<T> },
+		/// Change of ownership
+		OwnershipChanged { currency_id: CurrencyOf<T>, new_owner: AccountIdOf<T> },
+		/// Issuer and admin changed
+		ManagersChanged {
+			currency_id: CurrencyOf<T>,
+			new_admin: AccountIdOf<T>,
+			new_issuer: AccountIdOf<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Trying to register a new currency when id is in use
+		AlreadyCreated,
+		/// Trying to register a currency variant that is not ownable
+		NotOwnableCurrency,
+		/// Currency has not been created
+		NotCreated,
+		/// No permission to call the operation
+		NoPermission,
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	// The pallet's dispatchable functions.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create and take ownership of one CurrencyId
+		///
+		/// The creator will have full control of this pallelt's functions
+		/// regarding this currency
+		///
+		/// Parameters:
+		/// - `currency_id`: Currency id of the Token(u64) variant.
```

Comment:
Let's maybe not mention the `Token(u64)` here as it's not explicitly mentioned/used anywhere in this pallet.